### PR TITLE
CM-20592 - implement task timeout logic

### PR DESCRIPTION
**Pull request by MichalBor**

### CODE REVIEW

File: cli/utils/task_timer.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                 repeat_function: Optional[FunctionContext] = None,
+                 repeat_interval: Optional[int] = None):
+        Thread.__init__(self)
+        self._timeout = timeout
+        self._quit_function = quit_function
+        self._repeat_function = repeat_function
+        self._repeat_interval = repeat_interval
+        self.event = Event()
+
+    def run(self):
```
Comment (avishaiamiel) [inline cli/utils/task_timer.py:40]: Do we need validation over quit_function parameter?

File: cli/utils/task_timer.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    def stop(self):
+        self.event.set()
+
+    def _run_quit_function_on_timeout(self):
+        self.event.wait(self._timeout)
+        if not self.event.is_set():
+            self._call_quit_function()
+        self.stop()
+
+    def _run_repeat_function_until_timeout_and_quit_function_on_timeout(self):
```
Comment (avishaiamiel) [inline cli/utils/task_timer.py:59]: Nice!

**Review by avishaiamiel — CHANGES_REQUESTED** [review]
Really nice! Great job!

File: cli/utils/task_timer.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+from threading import Thread, Event
+from _thread import interrupt_main
+from typing import Optional, Callable, List, Dict, Type
+from types import TracebackType
+
+
+class FunctionContext:
+    FUNCTION: Callable
+    ARGS: Optional[List]
+    KWARGS: Optional[Dict]
```
[the remark this answers is not in the document]
Comment (MichalBor) [reply cli/utils/task_timer.py:10]: not mandatory, but I like it

File: cli/utils/task_timer.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    def __init__(self, function: Callable, args: Optional[List] = None, kwargs: Optional[Dict] = None):
+        self.FUNCTION = function
+        self.ARGS = args or []
+        self.KWARGS = kwargs or {}
```
[the remark this answers is not in the document]
Comment (MichalBor) [reply cli/utils/task_timer.py]: the ide did not like it. leave it as is

**Review by avishaiamiel — APPROVED** [review]

**Pull request merged**